Stratis.Bitcoin.Features.LightWallet/LightWalletSyncManager.cs

```diff
@@ -110,43 +110,79 @@ public void Stop()
                     while (this.chain.GetBlock(fork.HashBlock) == null)
                         fork = fork.Previous;
 
-                    Guard.Assert(fork.HashBlock == block.Header.HashPrevBlock);
+                    this.logger.LogInformation("Reorg detected, going back from '{0}/{1}' to '{2}/{3}'.", this.walletTip.HashBlock, this.walletTip.Height, fork.HashBlock, fork.Height);
+
                     this.walletManager.RemoveBlocks(fork);
-                    this.logger.LogWarning($"Reorg detected, wallet tip reverted back to Height = {fork.Height} hash = {fork.HashBlock}.");
+                    this.walletTip = fork;
+
+                    this.logger.LogTrace("Wallet tip set to '{0}'.", this.walletTip);
+                }
+
+                // The new tip can be ahead or behind the wallet
+                // If the new tip is ahead we try to bring the wallet up to the new tip.
+                // If the new tip is behind we just check the wallet and the tip are in the same chain.
+
+                if (newTip.Height > this.walletTip.Height)
+                {
+                    ChainedBlock findTip = newTip.FindAncestorOrSelf(this.walletTip.HashBlock);
+                    if (findTip == null)
+                    {
+                        this.logger.LogTrace("(-)[NEW_TIP_AHEAD_NOT_IN_WALLET]");
+                        return;
+                    }
+
+                    this.logger.LogTrace("Wallet tip '{0}/{1}' is behind the new tip '{2}/{3}'.", this.walletTip.HashBlock, this.walletTip.Height, newTip.HashBlock, newTip.HashBlock);
+
+                    // the wallet is falling behind we need to catch up
+                    this.logger.LogWarning("New tip '{0}/{1}' is too far in advance, put the puller back.", newTip.HashBlock, newTip.HashBlock);
+                    this.blockNotification.SyncFrom(this.walletTip.HashBlock);
+                    return;
                 }
                 else
                 {
-                    ChainedBlock incomingBlock = this.chain.GetBlock(block.GetHash());
-                    if (incomingBlock.Height > this.walletTip.Height)
+                    ChainedBlock findTip = this.walletTip.FindAncestorOrSelf(newTip.HashBlock);
+                    if (findTip == null)
                     {
-                        // the wallet is falling behind we need to catch up
-                        this.logger.LogWarning($"block received with height: {inBestChain.Height} and hash: {block.Header.GetHash()} is too far in advance. put the puller back.");
-                        this.blockNotification.SyncFrom(this.walletTip.HashBlock);
+                        this.logger.LogTrace("(-)[NEW_TIP_BEHIND_NOT_IN_WALLET]");
                         return;
                     }
+                    this.logger.LogTrace("Wallet tip '{0}/{1}' is ahead or equal to the new tip '{2}/{3}'.", this.walletTip.HashBlock, this.walletTip.Height, newTip.HashBlock, newTip.HashBlock);
```

Comment:
This can happen rarely (if the node crashed and consensus did not save its tip) it can also happen in the light wallet if you set the puller back in time to pull blocks again.

The ProcessesBlock is idempotent so no problem to processes a block again.

Reply:
I can take your word on it, I am not familiar with PB. It sounds dangerous on the first look, but if you know it is (or should be) safe, then OK, although I'd be slightly nervous about it.